### ISSUE TITLE
OBB probiou replace

### ISSUE THREAD

**Issue by djdll** [ultralytics/ultralytics#18119]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,
In OBB task
first: iou loss ：
```
class RotatedBboxLoss(BboxLoss):
    def forward(self, pred_dist, pred_bboxes, anchor_points, target_bboxes, target_scores, target_scores_sum, fg_mask):
        """IoU loss."""
        weight = target_scores.sum(-1)[fg_mask].unsqueeze(-1)
        iou, (pred_obb_angle, target_obb_angle) = probiou(pred_bboxes[fg_mask], target_bboxes[fg_mask],CIoU=False)
        loss_iou = ((1.0 - iou) * weight).sum() / target_scores_sum
```
print fg_mask.shape    : torch.Size([256, 4116])
print pred_bboxes.shape : torch.Size([256, 4116, 5])
print pred_bboxes[fg_mask].shape : torch.Size([5741, 5])
Because my understanding is that the loss is calculated by filtering all the boxes of a batch into an array through fg_mask. There is no batch dimension B ，I'm not sure I got that right `pred_bboxes[fg_mask]`

Now, I want to modify this loss, but the calculation requires a bach dimension, so I operate like this
`obb1.unsqueeze(0), obb2.unsqueeze(0)`
shape [5741, 5] -> [1, 5741, 5] like (B, N, 5) to calculating iou
but the loss I calculated in this case is very large
I don't think that's normal
```
        weight = target_scores.sum(-1)[fg_mask].unsqueeze(-1)
        iou, (pred_obb_angle, target_obb_angle) = probiou(pred_bboxes[fg_mask], target_bboxes[fg_mask],CIoU=False)
        loss_iou = ((1.0 - iou) * weight).sum() / target_scores_sum
```
I want to ask if it needs to be modified loss_iou = ((1.0 - iou) * weight).sum() / target_scores_sum
I think the main reason is that I don't understand pred_bboxes[fg_mask] or some details I don't know. Can you give me some suggestions. How to replace probiou for other calculation methods

**Comment by UltralyticsAssistant** [comment]
👋 Hello @djdll, thank you for your interest in Ultralytics 🚀!

We recommend a visit to the [Docs](https://docs.ultralytics.com) for new users where you can find many helpful resources, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. These resources may address some of your questions or provide insights into your implementation.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug the issue better. Ensure the example includes complete details about your setup and what behavior you are observing that's unexpected. This helps us investigate and provide a solution faster.

If this is a custom implementation ❓ Question, as it appears to be in your case, please provide more details (e.g., specifically what, in your loss computation, seems abnormal, any dataset snippets, or other contextual information). For general guidance, don't forget to check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/), as this might help improve your model modifications.

Here are a few additional recommendations:

### Upgrade
Make sure you're using the latest version of the `ultralytics` package containing all bug fixes and updates. You can upgrade with the following:

```bash
pip install -U ultralytics
```

### Environments

To ensure compatibility and proper execution of your tasks, make use of one of the following well-tested environments:

- **Notebooks** with free GPU:  
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See our [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/).
- **Amazon** Deep Learning AMI. See our [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker Image**. Check out our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

### Community Support

For real-time chat, consider joining our [Discord](https://discord.com/invite/ultralytics) 🎧. For deeper discussions on custom implementations like yours, try our [Discourse](https://community.ultralytics.com). You can also share and discuss ideas with others in our [Subreddit](https://reddit.com/r/Ultralytics).

### CI Status
Lastly, note that our automated [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests run daily and on every code commit to ensure all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) are working correctly.  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a> If the badge is green, all internal tests are currently passing! ✅

This is an automated response to help direct your attention to useful resources, but don't worry—a member of the Ultralytics team will follow up shortly to assist you further. 😊

**Comment by djdll** [comment]
loss_iou = ((1.0 - iou.mean()) * weight).sum() / target_scores_sum
add iou.mean() 
It seems that the data size is normal, maybe the problem with the data before is that the average is not taken

**Comment by glenn-jocher** [comment]
@djdll you're correct that adding `iou.mean()` ensures proper scaling for the loss calculation when the data size is normal. Averaging the IoU values can help stabilize the loss computation across batches. Ensure consistency in your implementation and verify the shapes of all tensors involved to avoid mismatches. For more details on `probiou`, you can refer to the [documentation](https://docs.ultralytics.com/reference/utils/metrics/).

**Comment by djdll** [comment]
@glenn-jocher Thank you very much. I successfully replaced probiou with rotatediou for the experiment through the documentation you provided, the understanding of the code and debugging.

**Comment by glenn-jocher** [comment]
You're welcome! Great to hear that you successfully implemented `rotatediou`. Kudos to your efforts and experimentation—best of luck with your project!

**Comment by djdll** [comment]
@glenn-jocher 
During training, it was found that when amp was turned on, the loss would be nan, and then the loss would be normal when I turned off amp. Why does this happen when the loss  is relatively large in the early training period
amp=True
1/1000      13.1G        nan      6.253
amp=False
1/1000      26.5G      5.054      7.141
